xmppserver/src/main/java/org/jivesoftware/openfire/muc/spi/MUCPersistenceManager.java

```diff
@@ -796,23 +797,29 @@ public static void loadHistory(@Nonnull final MUCRoom room, final int maxNumber)
                     Log.debug("Unable to skip to the last {} rows of the result set.", maxNumber, e);
                 }
 
+                final List<Message> oldMessages = new LinkedList<>();
                 while (rs.next()) {
                     String senderJID = rs.getString("sender");
                     String nickname = rs.getString("nickname");
                     Date sentDate = new Date(Long.parseLong(rs.getString("logTime").trim()));
                     String subject = rs.getString("subject");
                     String body = rs.getString("body");
                     String stanza = rs.getString("stanza");
-                    room.getRoomHistory().addOldMessage(senderJID, nickname, sentDate, subject, body, stanza);
+                    oldMessages.add(room.getRoomHistory().parseHistoricMessage(senderJID, nickname, sentDate, subject, body, stanza));
+                }
+
+                if (!oldMessages.isEmpty()) {
+                    room.getRoomHistory().addOldMessages(oldMessages);
```

Comment:
Any memory exhaustion concerns here? 
For a short while here, there's double the committed memory.
You'll have seen some esoteric setups, as well as Ignite, which has significant history. You might judge this to be low risk.

Reply:
The data is limited to that of one room. I don't see how that would be that much of a problem.

Reply:
True. 100K msgs at 100b per msg would only be a temporary 9MB spike.